src/ImageSharp/Formats/Png/PngDecoderCore.cs

```diff
@@ -1174,6 +1178,76 @@ private bool TryReadLegacyExifTextChunk(ImageMetadata metadata, string data)
             return true;
         }
 
+        /// <summary>
+        /// Reads the color profile chunk. The data is stored similar to the zTXt chunk.
+        /// </summary>
+        /// <param name="metadata">The metadata.</param>
+        /// <param name="data">The bytes containing the profile.</param>
+        private void ReadColorProfileChunk(ImageMetadata metadata, ReadOnlySpan<byte> data)
+        {
+            int zeroIndex = data.IndexOf((byte)0);
+            if (zeroIndex is < PngConstants.MinTextKeywordLength or > PngConstants.MaxTextKeywordLength)
+            {
+                return;
+            }
+
+            byte compressionMethod = data[zeroIndex + 1];
+            if (compressionMethod != 0)
+            {
+                // Only compression method 0 is supported (zlib datastream with deflate compression).
+                return;
+            }
+
+            ReadOnlySpan<byte> keywordBytes = data.Slice(0, zeroIndex);
+            if (!this.TryReadTextKeyword(keywordBytes, out string name))
+            {
+                return;
+            }
+
+            ReadOnlySpan<byte> compressedData = data.Slice(zeroIndex + 2);
+
+            if (this.TryUncompressZlibData(compressedData, out byte[] iccpProfileBytes))
+            {
+                metadata.IccProfile = new IccProfile(iccpProfileBytes);
+            }
+        }
+
+        /// <summary>
+        /// Tries to un-compress zlib compressed data.
+        /// </summary>
+        /// <param name="compressedData">The compressed data.</param>
+        /// <param name="uncompressedBytesArray">The uncompressed bytes array.</param>
+        /// <returns>True, if de-compressing was successful.</returns>
+        private unsafe bool TryUncompressZlibData(ReadOnlySpan<byte> compressedData, out byte[] uncompressedBytesArray)
+        {
+            fixed (byte* compressedDataBase = compressedData)
+            {
+                using (var memoryStream = new UnmanagedMemoryStream(compressedDataBase, compressedData.Length))
+                using (var bufferedStream = new BufferedReadStream(this.Configuration, memoryStream))
+                using (var inflateStream = new ZlibInflateStream(bufferedStream))
+                {
+                    if (!inflateStream.AllocateNewBytes(compressedData.Length, false))
+                    {
+                        uncompressedBytesArray = Array.Empty<byte>();
+                        return false;
+                    }
+
+                    var uncompressedBytes = new List<byte>(compressedData.Length);
+
+                    // Note: this uses a buffer which is only 4 bytes long to read the stream, maybe allocating a larger buffer makes sense here.
```

Comment:
Should we use the memory allocator to to create a bigger buffer to read the uncompressed data into here? using the scratch buffer feels like it could be very inefficient.

Reply:
Yeah, lot's of small allocations. I would use the allocator to allocate a buffer of length `Configuration.StreamProcessingBufferSize`.

Reply:
I was thinking about this for a while now, but if we always use the allocator this could be also inefficient, if we have alot of small compressed text string's.

Maybe something like this:
```
Span<byte> destBuffer = this.buffer.AsSpan();
using IMemoryOwner<byte> foo = compressedData.Length > 1024 ? this.memoryAllocator.Allocate<byte>(1024) : null;
if (compressedData.Length > 1024)
{
   destBuffer = foo.GetSpan();
}
```

For the the buffer to uncomress into?

Reply:
```c#
uncompressedBytes.AddRange(this.buffer.AsSpan(0, bytesRead).ToArray());
```

It's actually this line that concerns me the most. Zlib compression is pretty good and you could end up with a large number of `ToArray()` calls. That's why I favour a much larger array and simply slicing it. Pooling is pretty cheap.

Reply:
I dont see how this `ToArray` call can be avoided, though. We cannot pass a Span to `AddRange`, see [dotnet1530](https://github.com/dotnet/runtime/issues/1530)

Reply:
Maybe `MemoryStream`?

Can be used with spans and can expand like a list.
No virt calls as we would work with direct type.
Should be relatively fast as it uses `Buffer.InternalBlockCopy` for big buffers.

Reply:
@br3aker good idea, that will work: a0e38c87